### PR TITLE
Technical specifications

### PR DESCRIPTION
**Pull request by Malo-Archimbaud**
Approval required for the technical specifications

### CODE REVIEW

File: docs/Technical.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+In order to ease the use of the website by users, we will have cookies in which we will store the adress typed, that we will be activited only if the user agrees to it. Thanks to this, the user will not have to type his adress each time is going on the website.
+The website will be accessible through any web browser. 
+
+### **Back-end**
+
+The website will be deployed on an nginx server. The nginx server will take care of redirecting the adress type to the database, process the calculations to get the remaining time until Christmas and sending back the info to the front-end.
+
+### **Database**
+
+For our database, we will use the data from [geofabrik](http://download.geofabrik.de). This dataset is quite heavy but it allows us to have a great precision, and it has also regular updates. The data will be stored in a PostgreSQL database.
```
Comment (Guillaume-Riviere) [inline docs/Technical.md:89]: PostGre ? We use Photon i think

File: docs/Technical.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+In the case where this database is not fast enough or if is manipulation is too complicated, we could use data from this [dataset](http://www.geonames.org). This dataset is less precise because it only take in account the city and not the complete adress, but it is less heavier and easier to manipulate than the data from openstreetmap.
```
Comment (Guillaume-Riviere) [inline docs/Technical.md:90]: Its manipulation

File: docs/Technical.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+### Security threats
+
+- Someone could get the address or the GPS coordinates of the users.
+
+- A lot of people could try to connect at the same time at the service.
+
+### Security measures
+
+- The data that a user will type on the website will not be registered anywhere to be sure that nobody can access a user's location.
```
Comment (Guillaume-Riviere) [inline docs/Technical.md:114]: The adress will be in Cookies if the user accept it

File: docs/Technical.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+- Thanks to kubernetes service from Azure, the website will auto-scale to adapt to the inflow.
+
+### Privacy considerations
+
+- Only the adress typed could be stored in cookies, only if the user wants to
+
+### Risks
+
+- The connection between azure and the website could crash, leading to an incapacity to access the website.
+
```
Comment (Guillaume-Riviere) [inline docs/Technical.md:125]: What if we run out of credits on Azure ?

File: docs/Technical.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+### **3.c Error messages**
+
+| Errors | Meanings |
+| ------ | -------- |
+| "It seems that the adress you typed is wrong. You may have done a mistake" | The adress is not typed in a valid format. The user needs to type it again |
+| "I can not find you. It seems you did not activite your localisation" | The user tried to use the geolocalisation feature but he did not allow his browser to have access to his localisation |
+| | |
```
Comment (Guillaume-Riviere) [inline docs/Technical.md:132]: If on MAC, you also need to allow your browser to access your localisation in Mac's privacy settings

File: docs/Technical.md (hunk 1 above)
Comment (Guillaume-Riviere) [reply docs/Technical.md:89]: And also Why photon or Postgre

File: docs/Technical.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+The proposed solution is to create a service which will be able to, thanks to an address, tell with a precision of one second when exactly Santa is coming to deliver gifts.
+A countdown will be displayed according to the user's location which shows how much time is left until December the 25th, midnight of the sun.
+The solution will be deployed thanks to Kubernetes service on Azure.
+
+### **Front-end**
+
+On the website, a search bar will be displayed in which the user has to enter the adress he wants to know when Santa will arrive. Then, if the adress typed is correct, a countdown until Santa arrives at the midnight of the sun at this adress will be displayed.
+To calculate when Santa will arrive, we will add or subtract time depending on if the adress if at a negative or postive according to GMT (greenwich meridian). The countdown will be displayed in an DD/HH/MM/SS.
+A Reindeer wiil be displayed on the bottom left of the page. It will tell the user what to type in the searchbar, or warn the user if what they typed in is in a valid format (For more info go see the [error messages](#3c-error-messages)).
```
[the remark this answers is not in the document]
Comment (Malo-Archimbaud) [reply docs/Technical.md:79]: I just checked on 4 different websites and it seems that I am right. Deer is more for 'cerf', and reindeer is used for both 'caribou' and 'renne'.

File: docs/Technical.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
+# Technical Specification
+
+## Author(s)
+
+- Malo Archimbaud
+
+## Team
+
+| Role | Member |
```
[the remark this answers is not in the document]
Comment (Malo-Archimbaud) [reply docs/Technical.md:9]: done

File: docs/Technical.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+- [2. Solutions](#2-solutions)
+- [3. Further Considerations](#3-further-considerations)
+- [4. References](#4-references)
+
+## 1. Introduction
+
+### **1.a Overview**
+
+Our goal is to create a service which will be able to tell when Santa will come to deliver our gifts at our address to the second. Santa will deliver the gifts at midnight sharp depending on the sun. The website will be deployed thanks to Docker and Kubernetes. The project will be accessible through this [website](https://xmas.algosup.com)
+The problem is worth solving because there is no solution available now to know when Santa will come with a high degree of accuracy.
```
[the remark this answers is not in the document]
Comment (Malo-Archimbaud) [reply docs/Technical.md]: done

File: docs/Technical.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+### **1.b Glossary**
+
+| Term | Definition |
+| ---- | ---------- |
+| Midnight of the sun | This expression means that we want to know the exact moment when it is midnight regardless of timezones, but rather by calculate with the geocoordinates |
+| Docker | Docker is a platform which allows you to build, share and run applications on any server or platform. The app created is build in a container, which allows a faster deployment |
+| Kubernetes | Kubernetes is a system for automating deployment, scaling and management of containerized application |
+| Azure | Azure is the web sevice from Microsoft. Create kubernetes service, host a database and host your website are some of all the tools available |
+| Nginx (pronouced engine-x) | Nginx is a web server on which you can host a website and manage load |
```
[the remark this answers is not in the document]
Comment (Malo-Archimbaud) [reply docs/Technical.md]: done

File: docs/Technical.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+The solution will be deployed thanks to Kubernetes service on Azure.
+
+### **Front-end**
+
+On the website, a search bar will be displayed in which the user has to enter the adress he wants to know when Santa will arrive. Then, if the adress typed is correct, a countdown until Santa arrives at the midnight of the sun at this adress will be displayed.
+To calculate when Santa will arrive, we will add or subtract time depending on if the adress if at a negative or postive according to GMT (greenwich meridian). The countdown will be displayed in an DD/HH/MM/SS.
+A Reindeer wiil be displayed on the bottom left of the page. It will tell the user what to type in the searchbar, or warn the user if what they typed in is in a valid format (For more info go see the [error messages](#3c-error-messages)).
+In order to ease the use of the website by users, we will have cookies in which we will store the adress typed, that we will be activited only if the user agrees to it. Thanks to this, the user will not have to type his adress each time is going on the website.
```
[the remark this answers is not in the document]
Comment (Malo-Archimbaud) [reply docs/Technical.md]: done

File: docs/Technical.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+The website will be accessible through any web browser. 
+
+### **Back-end**
+
+The website will be deployed on an nginx server. The nginx server will take care of redirecting the adress type to the database, process the calculations to get the remaining time until Christmas and sending back the info to the front-end.
```
[the remark this answers is not in the document]
Comment (Malo-Archimbaud) [reply docs/Technical.md]: done

File: docs/Technical.md (hunk 1 above)
Comment (Malo-Archimbaud) [reply docs/Technical.md]: done

File: docs/Technical.md (hunk 2 above)
Comment (Malo-Archimbaud) [reply docs/Technical.md]: done

File: docs/Technical.md (hunk 3 above)
Comment (Malo-Archimbaud) [reply docs/Technical.md]: done

File: docs/Technical.md (hunk 4 above)
Comment (Malo-Archimbaud) [reply docs/Technical.md:125]: done

File: docs/Technical.md (hunk 5 above)
Comment (Malo-Archimbaud) [reply docs/Technical.md:134]: done

**Comment by Malo-Archimbaud** [comment]
I think I corrected everything that you said for now

File: docs/Technical.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Guillaume-Riviere) [reply docs/Technical.md]: Okay then =D

File: docs/Technical.md [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+In order to ease the use of the website by users, we will have cookies in which we will store the adress typed, that will be activited only if the user agrees to it. Thanks to this, the user will not have to type his adress each time he is going on the website.
+The website will be accessible through any web browser.
+
+### **Back-end**
+
+The website will be deployed on an nginx server. The nginx server will take care of intercating with the database. We used nginx over apache because when we have to deal with peak loads and we need to have a fast website, nginx is better than apache.
+
+### **Database**
+
+For our database, we will use the data from [geofabrik](http://download.geofabrik.de). This dataset is quite heavy but it allows us to have a great precision, and it has also regular updates. The data will be stored in a photon database. Photon will allow us to store our data as we wnat and is better when dealing with huge amount of data.
```
Comment (Guillaume-Riviere) [inline docs/Technical.md:89]: we 'want'

**Review by Guillaume-Riviere — CHANGES_REQUESTED** [review]

**Review by Guillaume-Riviere — APPROVED** [review]
All errors have been corrected 
All request have been made

**Pull request merged**